### PR TITLE
[REEF-1062] Update TaskConfiguration to properly use Tang Injection

### PR DESCRIPTION
**Pull request by afchung**
JIRA:
  [REEF-1062](https://issues.apache.org/jira/browse/REEF-1062)

### CODE REVIEW

**Comment by afchung** [comment]
@jwang98052 Please have a look at this. Thanks!

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextManager.cs [hunk 1]
```diff
@@ -338,8 +339,9 @@ private void StartTask(StartTaskProto startTaskProto)
                         string.Format(CultureInfo.InvariantCulture, "Task expected context '{0}' but the active context has Id '{1}'", expectedContextId, currentActiveContext.Id));
                     Utilities.Diagnostics.Exceptions.Throw(e, LOGGER);
                 }
-                TaskConfiguration taskConfiguration = new TaskConfiguration(startTaskProto.configuration);
-                currentActiveContext.StartTask(taskConfiguration, expectedContextId, _heartBeatManager);
+                
+                var configuration = new AvroConfigurationSerializer().FromString(startTaskProto.configuration);
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextManager.cs]: AvroConfigurationSerializer can be injected in the constructor of this class without any additional settings.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs [hunk 2]
```diff
@@ -226,17 +226,30 @@ public void StartTask(TaskConfiguration taskConfiguration, string contextId, Hea
                     var e = new InvalidOperationException("Attempting to instantiate a child context on a context that is not the topmost active context.");
                     Utilities.Diagnostics.Exceptions.Throw(e, LOGGER);
                 }
+
+                var taskInjector = _contextInjector.ForkInjector(taskConfiguration);
+                var taskId = string.Empty;
+                try
+                {
+                    taskId = taskInjector.GetNamedInstance<TaskConfigurationOptions.Identifier, string>();
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs]: GetNamedInstance is still a hack. You should use task configuration to inject a TaskConfiguration object and make  TaskConfigurationOptions.Identifier as a named parameter of the class.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs [hunk 3]
```diff
@@ -226,17 +226,30 @@ public void StartTask(TaskConfiguration taskConfiguration, string contextId, Hea
...
+                }
+
+                LOGGER.Log(Level.Info, "Trying to inject task with configuration" + taskConfiguration);
+
                 try
                 {
-                    IInjector taskInjector = _contextInjector.ForkInjector(taskConfiguration.TangConfig);
-                    LOGGER.Log(Level.Info, "Trying to inject task with configuration" + taskConfiguration.ToString());
-                    TaskRuntime taskRuntime = new TaskRuntime(taskInjector, contextId, taskConfiguration.TaskId, heartBeatManager);
+                    var taskRuntime = new TaskRuntime(taskInjector, contextId, taskId, heartBeatManager);
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs]: why not inject it?

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/RootContextLauncher.cs [hunk 4]
```diff
@@ -39,14 +39,14 @@ internal sealed class RootContextLauncher
...
-        public RootContextLauncher(ContextConfiguration rootContextConfig, Optional<ServiceConfiguration> rootServiceConfig, Optional<TaskConfiguration> rootTaskConfig)
+        public RootContextLauncher(ContextConfiguration rootContextConfig, Optional<ServiceConfiguration> rootServiceConfig, Optional<IConfiguration> rootTaskConfig)
         {
             _rootContextConfiguration = rootContextConfig;
             _rootServiceInjector = InjectServices(rootServiceConfig);
             RootTaskConfig = rootTaskConfig;
         }
 
-        public Optional<TaskConfiguration> RootTaskConfig { get; set; }
+        public Optional<IConfiguration> RootTaskConfig { get; set; }
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/RootContextLauncher.cs]: You might want of recover TaskConfiguration and make it injectable.

File: lang/cs/Org.Apache.REEF.Common/Tasks/TaskConfiguration.cs [hunk 5]
```diff
@@ -30,11 +26,8 @@
 
 namespace Org.Apache.REEF.Common.Tasks
 {
-    public class TaskConfiguration : ConfigurationModuleBuilder
+    public sealed class TaskConfiguration : ConfigurationModuleBuilder
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Tasks/TaskConfiguration.cs]: If you keep the original TaskConfiguration, I would suggest to change this one as TaskConfigurationModule. We should always have a separate class for configuration modules.

**Comment by jwang98052** [comment]
I did an initial pass. 
A general comment is we need to make all of them injectable, including TaskConfiguration, ContextRuntime and TaskRuntime. 
Also we don't have unit test for those. We should add unit test cases for all of those.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/RootContextLauncher.cs (hunk 4 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/RootContextLauncher.cs]: I'm not sure I understand, why do we need `TaskConfiguration`?

**Comment by afchung** [comment]
@jwang98052 `TaskRuntime` cannot be injected since `ContextConfiguration` is in the `Driver` project instead of the `Common` project. We will need to deprecate what is in the `Driver` project before we can move on with the rewrite using injection.

**Comment by afchung** [comment]
I'll close this first, since we cannot make progress before the `ContextConfiguration` problem is fixed.

**Comment by afchung** [comment]
@jwang98052 I've force pushed to this branch, so I'll need to open a new PR for this. Thanks!

**Pull request closed without merging**